add code to run strain profiling on transcriptomes and plot the results

taylorreiter opened the pull request:
This PR adds a snakefile to perform strain-level (genome) profiling of the sputum transcriptomes, an environment to control the rule-specific software, and a notebook to plot the resulting strain profiles.

+ `gather_for_strain.snakefile`: contains rules to generate `sourmash gather` profiles for each transcriptome of interest. Relies on two metadata data files, only one of which is currently available in this repo. I want to make sure the hogan lab metadata can be publicly viewable before adding it. The rule `rnaseq_sample_sourmash_sketch` checks if the fastq files exist locally before downloading/streaming them to sourmash sketches, so as not to duplicate downloads.
+ `envs/sourmash.yml`: contains the sourmash environment 
+ `notebooks/20220316_plot_gather_results.ipynb`: produces plots of the sourmash gather output, focusing on clinically important organisms

taylorreiter committed:
add snakefile for sourmash gather, env

taylorreiter committed:
add plotting code for gather

ajlee21 commented on gather_for_strain.snakefile:
```diff
@@ -0,0 +1,120 @@
+    inputs/raw/sra/{srx}.fastq.gz. If that file does exists, this rule will
+    sketch from that file. If it does not exists, this rule will streaming download
+    the SRA accession and pipe it to the sketch command, that way the raw data 
+    doesn't have to be written to disk. Because of this rule operating based on
```
Nice note. Not sure if you think it warrants making this last sentence its own line to say something like "Warning: Because...."? Totally up to you

taylorreiter replied:
```suggestion
    doesn't have to be written to disk. 
    Warning: Because of this rule operating based on
```

ajlee21 commented on gather_for_strain.snakefile:
```diff
@@ -0,0 +1,120 @@
+    '''
+
+
+###########################################################################
```
Very helpful having these section headers!

ajlee21 commented on gather_for_strain.snakefile:
```diff
@@ -0,0 +1,120 @@
+    output: "outputs/rnaseq_sourmash_sketch/{srx}.sig"
+    resources:
+        mem_mb = lambda wildcards, attempt: attempt * 2000 ,
+    threads: 1
```
If I wanted to parallelize this, would I need to update the number of threads here? Is there any other changes I'd need to do?

taylorreiter replied:
`threads` controls the number of threads that e.g. slurm will give to the rule when it's running on a cluster. When specified this way, it also becomes a python variable that can be called within e.g. the shell command. If the code is parallelizable with a flag (like `-t`), then you can write `-t {threads}` and it will scoop in the number of threads based on what's specified in the rule. Since this is also given to slurm by snakemake, it's works really nicely for code that is already parallelize-able. It would figure out how to parallelized code that doesn't know how to run in parallel. `sourmash sketch` can only run on one thread, so I only put a `1` here. 

The other handy thing snakemake will do is limit the total number of processes running at once. So when i go to run this on the command line, I would use `-j` to specify the maximum number of processes that can run at a time. so if i used `-j 16`, I'm telling snakemake that up to 16 threads can be started at a given time. So in this case, 16 sketch commands could launch because all of them only use 1 thread.

ajlee21 commented on gather_for_strain.snakefile:
```diff
@@ -0,0 +1,120 @@
+## Hogan lab samples
+######################################################################
+
+rule rnaseq_sourmash_sketch_spu:
```
I'm not sure if this is a misunderstanding of how snakefiles work, but is there a reason why you cannot have 1 rule to process both the public and Hogan lab data and just call the rule twice with different inputs?

taylorreiter replied:
no you're not misunderstanding, and it made me cringe to repeat this rule 🤣 however, i made the executive decision that it was okay because this specific workflow is pretty tiny. it had to be repeated the way i set it up because the input files are 1) in different input files and 2) specified by different metadata files. this is sort of just inherited from the fact that some of this workflow works with public data that has to be downloaded, and some of it is with private data that we assume is already on the hard disk.

ajlee21 reviewed (approved):
Nice plots. Just had a few questions to try to understand what the results are showing as a non-metagenomics person.

For future reviews, it is easier to review changes in .py files, so in addition to the jupyter notebooks you could also provide the converted .py files using this command: https://nbconvert.readthedocs.io/en/latest/usage.html

There is also a pre-commit that will automatically do this for you too: https://github.com/greenelab/onboarding/blob/master/extras/linter_install_tutorial.md

Some comments on the notebook:
1. It would be helpful to add some brief comments in sections for say: what is sourmash doing? what are the plots showing - abundance of strains within samples, but what do the metal vs non-metal mean? I assume that is the media.
2. What does "abundance-weighted unique fraction of sample" mean? The number of mapped reads to each strain?
3. So you have a few sputum samples with a lot of strain heterogeneity but then others with less than 5 (looking at your 2nd plot)?
4. What is the difference between the bottom 2 plots? Public samples vs Hogan lab samples? So Hogan lab samples are mainly composed of Pa but the public ones vary?

taylorreiter committed:
update note to warning

Co-authored-by: Alexandra Lee <alexjlee.21@gmail.com>

taylorreiter committed:
add context to viz nb

taylorreiter committed:
Merge branch 'gather' of github.com:taylorreiter/2022-cf-sputum into gather

taylorreiter commented:
> For future reviews, it is easier to review changes in .py files, so in addition to the jupyter notebooks you could also provide the converted .py files using this command: https://nbconvert.readthedocs.io/en/latest/usage.html
> 
> There is also a pre-commit that will automatically do this for you too: https://github.com/greenelab/onboarding/blob/master/extras/linter_install_tutorial.md
 
Thanks for the suggestion alex. I looked at the pre-commit workflow, as well as [a few other tools](https://github.com/deep5050/autopy-lot) and it looks like most pipelines don't have native support for R. Most of my notebooks will run on the irkernel, so i'm nervous to have multiple copies floating without automated guarantees that the `*R` or `.py` versions are copies. Do you have any thoughts on this?

> Some comments on the notebook:

Thank you for these comments! I added more descriptions to the notebooks so that all of these questions would be addressed.

> 1. It would be helpful to add some brief comments in sections for say: what is sourmash doing? what are the plots showing - abundance of strains within samples, but what do the metal vs non-metal mean? I assume that is the media.

Done! Added to the first description at the top of the nb.

> 2. What does "abundance-weighted unique fraction of sample" mean? The number of mapped reads to each strain?

Explained in the second section!
 
> 3. So you have a few sputum samples with a lot of strain heterogeneity but then others with less than 5 (looking at your 2nd plot)?

yes exactly

> 4. What is the difference between the bottom 2 plots? Public samples vs Hogan lab samples? So Hogan lab samples are mainly composed of Pa but the public ones vary?

yes -- it's mostly the difference between where the samples came from. added some context to help make this more clear!
